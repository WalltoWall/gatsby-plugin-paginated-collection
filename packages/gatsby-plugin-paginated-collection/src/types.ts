import {
  PluginOptions as GatsbyPluginOptions,
  NodeInput,
  CreatePagesArgs,
  Node,
} from 'gatsby'

export type ID = string

export enum NodeType {
  Collection = 'PaginatedCollection',
  Page = 'PaginatedCollectionPage',
}

export type NormalizedNode = Record<string, unknown>

export type PageNode = Node & PageNodeInput

export interface PageNodeInput extends NodeInput {
  collection: ID
  index: number
  nextPage: ID | undefined
  hasNextPage: boolean
  previousPage: ID | undefined
  hasPreviousPage: boolean
  nodeCount: number
  nodes: NormalizedNode[]
  internal: NodeInput['internal'] & {
    type: NodeType.Page
  }
}

export type CollectionNode = Node & CollectionNodeInput

export interface CollectionNodeInput extends NodeInput {
  name: string
  pageSize: number
  firstPageSize: number
  lastPageSize: number
  nodeCount: number
  pageCount: number
  pages: ID[]
  internal: NodeInput['internal'] & {
    type: NodeType.Collection
  }
}

export interface GraphQLResult {
  errors?: any
  data?: unknown
}

export type PluginOptions = Required<ProvidedPluginOptions>

export interface PluginConfig {
  resolve: string
  options?: Record<string, unknown>
}

export interface ProvidedPluginOptions extends GatsbyPluginOptions {
  name: string
  query: string
  pageSize?: number
  firstPageSize?: number
  normalizer: (input: GraphQLResult) => NormalizedNode[]
  plugins: PluginConfig[]
}

export interface Plugin {
  onPostCreateNodes?(
    gatsbyContext: CreatePagesArgs & { traceId: 'initial-createPages' },
    rootPluginOptions: PluginOptions,
    pluginOptions: unknown,
  ): void | Promise<void>
}
