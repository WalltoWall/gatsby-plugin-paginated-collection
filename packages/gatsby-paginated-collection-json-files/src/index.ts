import { promises as fs } from 'fs'
import path from 'path'
import {
  Plugin,
  PageNode,
  CollectionNode,
} from 'gatsby-plugin-paginated-collection'

const DEFAULT_PLUGIN_OPTIONS: Required<PluginOptions> = {
  path: 'paginated-collections',
  expand: [],
  filename: 'id',
}

export interface PluginOptions {
  path?: string
  expand?: ('nextPage' | 'previousPage' | 'collection')[]
  filename?: 'id' | 'index' | ((node: PageNode) => string)
}

export type ExpandedPageNode = PageNode & {
  nextPage?: PageNode & { nodes: undefined }
  previousPage?: PageNode & { nodes: undefined }
  collection: CollectionNode
}

export const onPostCreateNodes: Plugin['onPostCreateNodes'] = async (
  node,
  rawPluginOptions: PluginOptions,
  gatsbyContext,
  _rootPluginOptions,
) => {
  const { getNode, store } = gatsbyContext
  const { program } = store.getState()
  const pluginOptions: Required<PluginOptions> = {
    ...DEFAULT_PLUGIN_OPTIONS,
    ...rawPluginOptions,
  }

  const dir = path.resolve(program.directory, 'public', pluginOptions.path)
  await fs.mkdir(dir, { recursive: true })

  await Promise.all(
    node.pages.map(async pageId => {
      const page: ExpandedPageNode = getNode(pageId)

      if (pluginOptions.expand.includes('nextPage')) {
        const expansion = getNode(page.nextPage)
        if (expansion) page.nextPage = { ...expansion, nodes: undefined }
      }

      if (pluginOptions.expand.includes('previousPage')) {
        const expansion = getNode(page.previousPage)
        if (expansion) page.previousPage = { ...expansion, nodes: undefined }
      }

      if (pluginOptions.expand.includes('collection'))
        page.collection = getNode(page.collection)

      const fileBasename =
        typeof pluginOptions.filename === 'function'
          ? String(pluginOptions.filename(page))
          : page[pluginOptions.filename]

      await fs.writeFile(
        path.join(dir, fileBasename + '.json'),
        JSON.stringify(page),
      )
    }),
  )
}
