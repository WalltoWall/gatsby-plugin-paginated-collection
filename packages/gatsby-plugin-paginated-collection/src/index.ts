export { Plugin, NodeType, PageNode, CollectionNode } from './types'
