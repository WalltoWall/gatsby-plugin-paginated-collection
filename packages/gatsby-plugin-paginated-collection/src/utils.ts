/**
 * Formats a message suitable to display to a user.
 *
 * @param msg Message to format.
 *
 * @returns Formatted message.
 */
export const fmtMsg = (msg: string) =>
  `gatsby-plugin-paginated-collection - ${msg}`

/**
 * Creates an array of elements split into groups of the provided size.
 *
 * @param size - Length of each chunk.
 * @param array - The array to split.
 *
 * @returns An array containing chunked arrays. If array cannot be split evenly, the final chunk will be the remaining elements.
 */
export const chunk = <T>(size: number, arr: T[]): T[][] =>
  arr.reduce(
    (array, item, i) =>
      i % size === 0
        ? [...array, [item]]
        : [...array.slice(0, -1), [...array.slice(-1)[0], item]],
    [] as T[][],
  )
