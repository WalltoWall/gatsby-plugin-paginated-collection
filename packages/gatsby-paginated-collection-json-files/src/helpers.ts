import path from 'path'
import { promises as fs } from 'fs'
import { CollectionNode, PageNode } from 'gatsby-plugin-paginated-collection'
import { CreatePagesArgs } from 'gatsby'

export interface WritePaginatedCollectionJSONFilesArgs {
  node: CollectionNode
  directory: string
  expand?: ('nextPage' | 'previousPage' | 'collection')[]
  filename?: 'id' | 'index' | ((node: PageNode) => string)
  getNode: CreatePagesArgs['getNode']
}

export const writePaginatedCollectionJSONFiles = async (
  args: WritePaginatedCollectionJSONFilesArgs,
): Promise<void> => {
  const { node, directory, expand = [], filename = 'id', getNode } = args

  await fs.mkdir(directory, { recursive: true })

  await Promise.all(
    node.pages.map(async (pageId) => {
      const page: PageNode = getNode(pageId)
      const trimmedPage = {
        ...page,
        nextPage: expand.includes('nextPage')
          ? {
              ...getNode(page.nextPage),
              nodes: undefined,
              internal: undefined,
              children: undefined,
              parent: undefined,
            }
          : page.nextPage,
        previousPage: expand.includes('previousPage')
          ? {
              ...getNode(page.previousPage),
              nodes: undefined,
              internal: undefined,
              children: undefined,
              parent: undefined,
            }
          : page.previousPage,
        collection: expand.includes('collection')
          ? {
              ...getNode(page.collection),
              internal: undefined,
              children: undefined,
              parent: undefined,
            }
          : page.collection,
        internal: undefined,
        children: undefined,
        parent: undefined,
      }

      const fileBasename =
        typeof filename === 'function' ? String(filename(page)) : page[filename]

      await fs.writeFile(
        path.join(directory, fileBasename + '.json'),
        JSON.stringify(trimmedPage),
      )
    }),
  )
}
