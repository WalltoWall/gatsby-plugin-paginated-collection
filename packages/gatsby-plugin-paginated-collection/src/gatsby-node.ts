import {
  GatsbyNode,
  CreateSchemaCustomizationArgs,
  PluginOptions as GatsbyPluginOptions,
} from 'gatsby'

import { chunk, fmtMsg, createCollectionNode } from './utils'
import { ProvidedPluginOptions, PluginOptions, Plugin } from './types'
import { types } from './gqlTypes'

export const DEFAULT_PLUGIN_OPTIONS = {
  pageSize: 10,
}

export const createPages: GatsbyNode['createPages'] = (
  gatsbyContext,
  providedPluginOptions: ProvidedPluginOptions,
  callback,
) => {
  const pluginOptions: PluginOptions = {
    ...DEFAULT_PLUGIN_OPTIONS,
    ...providedPluginOptions,
    firstPageSize:
      providedPluginOptions.firstPageSize ??
      providedPluginOptions.pageSize ??
      DEFAULT_PLUGIN_OPTIONS.pageSize,
  }

  const { graphql, getNode, reporter } = gatsbyContext
  const { name, query, pageSize, normalizer, plugins } = pluginOptions

  const asyncFn = async () => {
    const queryResult = await graphql(query)

    if (queryResult.errors) {
      reporter.error(
        fmtMsg(
          `The provided GraphQL query for "${name}" contains errors. Pagination files will not be created.`,
        ),
        queryResult.errors[0],
      )
      return
    }

    const items = await Promise.resolve(normalizer(queryResult))

    if (!Array.isArray(items)) {
      reporter.error(
        fmtMsg(
          `The normalizer function for "${name}" did not return an array. Pagination files will not be created.`,
        ),
      )
      return
    }

    if (items.length < 1)
      reporter.warn(
        fmtMsg(
          `No items for "${name}" were returned. Pagination data will be created on an empty set.`,
        ),
      )

    const chunks = chunk(pageSize, items)
    const nodeId = createCollectionNode(chunks, gatsbyContext, pluginOptions)
    const node = getNode(nodeId)

    for (const plugin of plugins) {
      const requiredPlugin: Plugin = require(plugin.resolve)
      await requiredPlugin?.onPostCreateNodes?.(
        node,
        plugin.options,
        gatsbyContext,
        pluginOptions,
      )
    }
  }

  asyncFn().finally(() => callback && callback(null))
}

export const createSchemaCustomization: GatsbyNode['createSchemaCustomization'] = async (
  gatsbyContext: CreateSchemaCustomizationArgs,
  _pluginOptions: GatsbyPluginOptions,
) => {
  const { actions } = gatsbyContext
  const { createTypes } = actions

  createTypes(types)
}
