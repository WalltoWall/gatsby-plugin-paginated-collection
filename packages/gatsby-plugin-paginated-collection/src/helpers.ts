import { CreatePagesArgs } from 'gatsby'

import {
  NormalizedNode,
  NodeType,
  ID,
  PageNodeInput,
  CollectionNodeInput,
} from './types'
import { chunk } from './utils'

interface CreatePageNodeArgs {
  /** Nodes for the page. */
  chunk: NormalizedNode[]
  /** UUID for the page. Used as a pseudo cursor to identify the page. */
  id: ID
  /** Index of the page within the collection. */
  index: number
  /** Tuples Tuples of page UUIDs and nodes. */
  chunkTuples: [ID, NormalizedNode[]][]
  collectionId: ID
  /** Gatsby createPages context. */
  createNode: CreatePagesArgs['actions']['createNode']
  /** Options provided to the plugin by the user. */
  createContentDigest: CreatePagesArgs['createContentDigest']
}

/**
 * Creates a Gatsby Node representing a page in a collection of nodes.
 *
 * @param args Arguments to create the page node.
 *
 * @returns The created Gatsby Node.
 */
const createPageNode = (args: CreatePageNodeArgs): PageNodeInput => {
  const {
    chunk,
    id,
    index,
    chunkTuples,
    collectionId,
    createNode,
    createContentDigest,
  } = args

  const nextChunkTuple = chunkTuples[index + 1]
  const previousChunkTuple = chunkTuples[index - 1]

  const node: PageNodeInput = {
    id,
    collection: collectionId,
    index,
    nextPage: nextChunkTuple?.[0],
    hasNextPage: Boolean(nextChunkTuple),
    previousPage: previousChunkTuple?.[0],
    hasPreviousPage: Boolean(previousChunkTuple),
    nodeCount: chunk.length,
    nodes: chunk,
    internal: {
      type: NodeType.Page,
      contentDigest: createContentDigest(chunk),
    },
  }

  createNode(node, { name: 'gatsby-plugin-paginated-collection' })

  return node
}

interface CreatePaginatedCollectionNodesArgs {
  /** Array of items to paginate. */
  collection: NormalizedNode[]
  /** Name of the collection. */
  name: string
  /** The number of nodes to include in each page. */
  pageSize: number
  /** The action used to create nodes. */
  createNode: CreatePagesArgs['actions']['createNode']
  /** A helper function for creating node IDs. */
  createNodeId: CreatePagesArgs['createNodeId']
  /** A helper function for creating node content digests. */
  createContentDigest: CreatePagesArgs['createContentDigest']
}

/**
 * Creates a Gatsby Node and its child nodes representing a collection of pages
 * of nodes.
 *
 * @param args Arguments to create the collection nodes.
 *
 * @returns The created Gatsby Node.
 */
export const createPaginatedCollectionNodes = (
  args: CreatePaginatedCollectionNodesArgs,
) => {
  const {
    collection,
    name,
    pageSize,
    createNode,
    createNodeId,
    createContentDigest,
  } = args

  const id: string = createNodeId(`${NodeType.Collection} ${name}`)

  const chunks = chunk(pageSize, collection)
  const pageNodes = chunks
    .map<[ID, NormalizedNode[]]>((chunk, index) => [
      createNodeId(`${NodeType.Page} ${name} ${index}`),
      chunk,
    ])
    .map(([pageId, chunk], index, arr) =>
      createPageNode({
        chunk,
        id: pageId,
        index,
        chunkTuples: arr,
        collectionId: id,
        createNode,
        createContentDigest,
      }),
    )

  const nodeCount = pageNodes.reduce(
    (count, pageNode) => (count += pageNode.nodes.length),
    0,
  )

  const pageNodeIds = pageNodes.map(pageNode => pageNode.id)

  const node: CollectionNodeInput = {
    id,
    name,
    pageSize,
    firstPageSize: pageNodes[0].nodes.length,
    lastPageSize: pageNodes[pageNodes.length - 1].nodes.length,
    nodeCount,
    pageCount: pageNodes.length,
    pages: pageNodeIds,
    internal: {
      type: NodeType.Collection,
      contentDigest: createContentDigest(pageNodes),
    },
  }

  createNode(node, { name: 'gatsby-plugin-paginated-collection' })

  return node
}
