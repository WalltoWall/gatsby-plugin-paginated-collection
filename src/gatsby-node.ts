import { GatsbyNode } from 'gatsby'

import { chunk, fmtMsg, createCollectionNode } from './utils'
import { ProvidedPluginOptions, PluginOptions } from './types'

const DEFAULT_PLUGIN_OPTIONS = {
  pageSize: 10,
}

export const createPages: GatsbyNode['createPages'] = (
  gatsbyContext,
  providedPluginOptions: ProvidedPluginOptions,
  callback,
) => {
  const pluginOptions: PluginOptions = {
    ...DEFAULT_PLUGIN_OPTIONS,
    ...providedPluginOptions,
    firstPageSize:
      providedPluginOptions.firstPageSize ??
      providedPluginOptions.pageSize ??
      DEFAULT_PLUGIN_OPTIONS.pageSize,
  }

  const { graphql, reporter } = gatsbyContext
  const { name, query, pageSize, normalizer } = pluginOptions

  const asyncFn = async () => {
    const queryResult = await graphql(query)

    if (queryResult.errors) {
      reporter.error(
        fmtMsg(
          `The provided GraphQL query for "${name}" contains errors. Pagination files will not be created.`,
        ),
        queryResult.errors[0],
      )
      return
    }

    const items = await Promise.resolve(normalizer(queryResult))

    if (!Array.isArray(items)) {
      reporter.error(
        fmtMsg(
          `The normalizer function for "${name}" did not return an array. Pagination files will not be created.`,
        ),
      )
      return
    }

    if (items.length < 1)
      reporter.warn(
        fmtMsg(
          `No items for "${name}" were returned. Pagination data will be created on an empty set.`,
        ),
      )

    const chunks = chunk(pageSize, items)
    createCollectionNode(chunks, gatsbyContext, pluginOptions)
  }

  asyncFn().finally(() => callback && callback(null))
}
