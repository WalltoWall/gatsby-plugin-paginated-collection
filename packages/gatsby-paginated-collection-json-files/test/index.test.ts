import path from 'path'
import fs from 'fs'
import mockFs from 'mock-fs'
import { CreatePagesArgs } from 'gatsby'

import { onPostCreateNodes, PluginOptions } from '../src'
import {
  CollectionNode,
  NodeType,
} from 'gatsby-plugin-paginated-collection/src/types'

const MOCK_PROGRAM_DIRECTORY_PATH = '/__PROGRAM_DIRECTORY__/'

const rootPluginOptions = {
  name: 'name',
  query: 'query',
  pageSize: 1,
  firstPageSize: 1,
  normalizer: () => [],
  plugins: [{ resolve: 'gatsby-paginated-collection-json-files' }],
}

const nodes = [
  { id: 'node1', foo: 'bar' },
  { id: 'node2', foo: 'baz' },
]

const pages = [
  {
    id: 'page1',
    index: 0,
    nextPage: 'page2',
    nodes: [nodes[0]],
    collection: 'collection1',
    internal: {},
    children: [],
    parent: null,
  },
  {
    id: 'page2',
    index: 1,
    previousPage: 'page1',
    nodes: [nodes[1]],
    collection: 'collection1',
    internal: {},
    children: [],
    parent: null,
  },
]

const collections = [
  {
    id: 'collection1',
    pages: pages.map((p) => p.id),
    internal: {},
    children: [],
    parent: null,
  },
]

const mockCollectionNode: CollectionNode = {
  id: 'id',
  name: 'name',
  pageSize: rootPluginOptions.pageSize,
  firstPageSize: rootPluginOptions.firstPageSize,
  lastPageSize: pages[pages.length - 1].nodes.length,
  nodeCount: nodes.length,
  pageCount: collections[0].pages.length,
  pages: pages.map((p) => p.id),
  parent: 'parent',
  children: [],
  internal: {
    type: NodeType.Collection,
    contentDigest: 'contentDigest',
    owner: 'gatsby-plugin-paginated-collection',
  },
}

const mockActions = {
  deletePage: jest.fn(),
  createPage: jest.fn(),
  deleteNode: jest.fn(),
  deleteNodes: jest.fn(),
  createNode: jest.fn(),
  touchNode: jest.fn(),
  createNodeField: jest.fn(),
  createParentChildLink: jest.fn(),
  setWebpackConfig: jest.fn(),
  replaceWebpackConfig: jest.fn(),
  setBabelOptions: jest.fn(),
  setBabelPlugin: jest.fn(),
  setBabelPreset: jest.fn(),
  createJob: jest.fn(),
  createJobV2: jest.fn(),
  setJob: jest.fn(),
  endJob: jest.fn(),
  setPluginStatus: jest.fn(),
  createRedirect: jest.fn(),
  addThirdPartySchema: jest.fn(),
  createTypes: jest.fn(),
  createFieldExtension: jest.fn(),
}

const mockGatsbyContext: CreatePagesArgs & {
  traceId: 'initial-createPages'
} = {
  pathPrefix: 'pathPrefix',
  boundActionCreators: mockActions,
  actions: mockActions,
  loadNodeContent: jest.fn(),
  store: {
    dispatch: jest.fn(),
    subscribe: jest.fn(),
    getState: jest
      .fn()
      .mockReturnValue({ program: { directory: MOCK_PROGRAM_DIRECTORY_PATH } }),
    replaceReducer: jest.fn(),
  },
  emitter: {
    addListener: jest.fn(),
    on: jest.fn(),
    once: jest.fn(),
    prependListener: jest.fn(),
    prependOnceListener: jest.fn(),
    removeListener: jest.fn(),
    off: jest.fn(),
    removeAllListeners: jest.fn(),
    setMaxListeners: jest.fn(),
    getMaxListeners: jest.fn(),
    listeners: jest.fn(),
    rawListeners: jest.fn(),
    emit: jest.fn(),
    eventNames: jest.fn(),
    listenerCount: jest.fn(),
  },
  getNodes: jest.fn(),
  getNode: jest
    .fn()
    .mockImplementation((id: string) =>
      [...collections, ...pages, ...nodes].find((p) => p.id === id),
    ),
  getNodesByType: jest.fn(),
  hasNodeChanged: jest.fn(),
  reporter: {
    stripIndent: jest.fn(),
    format: jest.fn(),
    setVerbose: jest.fn(),
    setNoColor: jest.fn(),
    panic: jest.fn(),
    panicOnBuild: jest.fn(),
    error: jest.fn(),
    uptime: jest.fn(),
    success: jest.fn(),
    verbose: jest.fn(),
    info: jest.fn(),
    warn: jest.fn(),
    log: jest.fn(),
    activityTimer: jest.fn(),
    createProgress: jest.fn(),
  },
  getNodeAndSavePathDependency: jest.fn(),
  cache: {
    getAndPassUp: jest.fn(),
    wrap: jest.fn(),
    set: jest.fn(),
    mset: jest.fn(),
    get: jest.fn(),
    mget: jest.fn(),
    del: jest.fn(),
    reset: jest.fn(),
  },
  createNodeId: jest.fn().mockReturnValue('createNodeId'),
  createContentDigest: jest.fn().mockReturnValue('createContentDigest'),
  tracing: {
    tracer: {},
    parentSpan: {},
    startSpan: jest.fn(),
  },
  schema: {
    buildObjectType: jest.fn(),
    buildUnionType: jest.fn(),
    buildInterfaceType: jest.fn(),
    buildInputObjectType: jest.fn(),
    buildEnumType: jest.fn(),
    buildScalarType: jest.fn(),
  },
  traceId: 'initial-createPages',
  waitForCascadingActions: false,
  parentSpan: {},
  graphql: jest.fn(),
}

beforeEach(() => {
  jest.clearAllMocks()
  mockFs({ MOCK_PROGRAM_DIRECTORY_PATH: {} })
})
afterEach(() => mockFs.restore())

describe('onPostCreateNodes', () => {
  test('creates files', async () => {
    await onPostCreateNodes!(
      mockCollectionNode,
      {},
      mockGatsbyContext,
      rootPluginOptions,
    )

    const dir = path.join(
      MOCK_PROGRAM_DIRECTORY_PATH,
      'public',
      'paginated-collections',
    )
    const filenames = fs.readdirSync(dir)
    const files = filenames.map((filename) =>
      fs.readFileSync(path.join(dir, filename), 'utf-8'),
    )
    mockFs.restore()

    expect(filenames).toEqual(Object.values(pages).map((p) => `${p.id}.json`))
    expect(files).toMatchSnapshot()
  })

  test('writes to configured directory', async () => {
    await onPostCreateNodes!(
      mockCollectionNode,
      { path: 'new-path' },
      mockGatsbyContext,
      rootPluginOptions,
    )

    const dir = path.join(MOCK_PROGRAM_DIRECTORY_PATH, 'public', 'new-path')
    const filenames = fs.readdirSync(dir)
    expect(filenames).toEqual(Object.values(pages).map((p) => `${p.id}.json`))
  })

  describe('expands fields', () => {
    test('nextPage', async () => {
      await onPostCreateNodes!(
        mockCollectionNode,
        { expand: ['nextPage'] },
        mockGatsbyContext,
        rootPluginOptions,
      )

      const dir = path.join(
        MOCK_PROGRAM_DIRECTORY_PATH,
        'public',
        'paginated-collections',
      )
      const filenames = fs.readdirSync(dir)
      const file = JSON.parse(
        fs.readFileSync(path.join(dir, filenames[0]), 'utf-8'),
      )
      expect(file.nextPage).toEqual({
        ...pages[1],
        nodes: undefined,
        internal: undefined,
        parent: undefined,
        children: undefined,
      })
    })

    test('previousPage', async () => {
      await onPostCreateNodes!(
        mockCollectionNode,
        { expand: ['previousPage'] },
        mockGatsbyContext,
        rootPluginOptions,
      )

      const dir = path.join(
        MOCK_PROGRAM_DIRECTORY_PATH,
        'public',
        'paginated-collections',
      )
      const filenames = fs.readdirSync(dir)
      const file = JSON.parse(
        fs.readFileSync(path.join(dir, filenames[1]), 'utf-8'),
      )
      expect(file.previousPage).toEqual({
        ...pages[0],
        nodes: undefined,
        internal: undefined,
        parent: undefined,
        children: undefined,
      })
    })

    test('collection', async () => {
      await onPostCreateNodes!(
        mockCollectionNode,
        { expand: ['collection'] },
        mockGatsbyContext,
        rootPluginOptions,
      )

      const dir = path.join(
        MOCK_PROGRAM_DIRECTORY_PATH,
        'public',
        'paginated-collections',
      )
      const filenames = fs.readdirSync(dir)
      const file = JSON.parse(
        fs.readFileSync(path.join(dir, filenames[0]), 'utf-8'),
      )
      expect(file.collection).toEqual({
        ...collections[0],
        internal: undefined,
        parent: undefined,
        children: undefined,
      })
    })
  })

  describe('uses correct filename', () => {
    test('id by default', async () => {
      await onPostCreateNodes!(
        mockCollectionNode,
        {},
        mockGatsbyContext,
        rootPluginOptions,
      )

      const dir = path.join(
        MOCK_PROGRAM_DIRECTORY_PATH,
        'public',
        'paginated-collections',
      )
      const filenames = fs.readdirSync(dir)
      mockFs.restore()

      expect(filenames).toEqual(Object.values(pages).map((p) => `${p.id}.json`))
    })

    test('determined by property name', async () => {
      await onPostCreateNodes!(
        mockCollectionNode,
        { filename: 'index' },
        mockGatsbyContext,
        rootPluginOptions,
      )

      const dir = path.join(
        MOCK_PROGRAM_DIRECTORY_PATH,
        'public',
        'paginated-collections',
      )
      const filenames = fs.readdirSync(dir)
      mockFs.restore()

      expect(filenames).toEqual(
        Object.values(pages).map((p) => `${p.index}.json`),
      )
    })

    test('determined by function', async () => {
      await onPostCreateNodes!(
        mockCollectionNode,
        { filename: (node) => node.index.toString() } as PluginOptions,
        mockGatsbyContext,
        rootPluginOptions,
      )

      const dir = path.join(
        MOCK_PROGRAM_DIRECTORY_PATH,
        'public',
        'paginated-collections',
      )
      const filenames = fs.readdirSync(dir)
      mockFs.restore()

      expect(filenames).toEqual(
        Object.values(pages).map((p) => `${p.index}.json`),
      )
    })
  })
})
