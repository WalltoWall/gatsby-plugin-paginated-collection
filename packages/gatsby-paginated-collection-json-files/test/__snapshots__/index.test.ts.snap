// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`onPostCreateNodes creates files 1`] = `
Array [
  "{\\"id\\":\\"page1\\",\\"index\\":0,\\"nextPage\\":\\"page2\\",\\"nodes\\":[{\\"id\\":\\"node1\\",\\"foo\\":\\"bar\\"}],\\"collection\\":\\"collection1\\"}",
  "{\\"id\\":\\"page2\\",\\"index\\":1,\\"previousPage\\":\\"page1\\",\\"nodes\\":[{\\"id\\":\\"node2\\",\\"foo\\":\\"baz\\"}],\\"collection\\":\\"collection1\\"}",
]
`;
